File: Assets/Typescript~/src/Shared/Util/Timer.ts
import { Signal } from "./Signal";

export const OnUpdate = new Signal<[deltaTime: number]>().WithYieldTracking(false);
export const OnLateUpdate = new Signal<[deltaTime: number]>().WithYieldTracking(false);
export const OnFixedUpdate = new Signal<[fixedDeltaTime: number]>().WithYieldTracking(false);
export const OnTick = new Signal<void>().WithYieldTracking(false);

export function SetTimeout<T extends unknown[]>(duration: number, callback: (...args: T) => void, ...args: T) {
	const triggerTime = Time.time + duration;
	const disconnect = OnUpdate.Connect(() => {
		if (Time.time >= triggerTime) {
			disconnect();
			callback(...args);
		}
	});
	return disconnect;
}

export function SetInterval(interval: number, callback: Callback, immediate?: boolean) {
	if (immediate) {
		callback();
	}
	let nextTriggerTime = Time.time + interval;
	const disconnect = OnUpdate.Connect(() => {
		const now = Time.time;
		if (now >= nextTriggerTime) {
			nextTriggerTime = now + interval;
			callback();
		}
	});
	return disconnect;
}

File: Assets/Bundles/Shared/Resources/TS/Util/Timer.lua
-- Compiled with unity-ts v2.1.0-75
local Signal = require("Shared/TS/Util/Signal").Signal
local OnUpdate = Signal.new():WithYieldTracking(false)
local OnLateUpdate = Signal.new():WithYieldTracking(false)
local OnFixedUpdate = Signal.new():WithYieldTracking(false)
local OnTick = Signal.new():WithYieldTracking(false)
local function SetTimeout(duration, callback, ...)
	local args = { ... }
	local triggerTime = Time.time + duration
	local disconnect
	disconnect = OnUpdate:Connect(function()
		if Time.time >= triggerTime then
			disconnect()
			callback(unpack(args))
		end
	end)
	return disconnect
end
local function SetInterval(interval, callback, immediate)
	if immediate then
		callback()
	end
	local nextTriggerTime = Time.time + interval
	local disconnect = OnUpdate:Connect(function()
		local now = Time.time
		if now >= nextTriggerTime then
			nextTriggerTime = now + interval
			callback()
		end
	end)
	return disconnect
end
return {
	SetTimeout = SetTimeout,
	SetInterval = SetInterval,
	OnUpdate = OnUpdate,
	OnLateUpdate = OnLateUpdate,
	OnFixedUpdate = OnFixedUpdate,
	OnTick = OnTick,
}
-- ----------------------------------
-- ----------------------------------
